progs/jubjub.py

```diff
@@ -219,3 +219,41 @@ def __str__(self):
 
     async def double(self):
         return self
+
+
+async def share_mul(context: Mpc, bs: list, p: Point) -> SharedPoint:
+    """
+    The multiplication of the share of a field element and a point
+    e.g. [x]P -> [X], where P is a point on the given elliptic curve
+    x is the bitwise shared value,
+    starting from the least significant bit.
+
+    NOTE: This is an affine version.
+    bs := [[b0], [b1], ... [bK]], then bs * P can be broken down into
+    [b0] * (2^0 * P) + [b1] * (2^1 * P) .... + [bK] * (2^K * P)
+
+    For each term [bi] * (2^i * P), we compute its x, y coordinates seperately.
+    Let P2i := (2^i * P), and we have identity = (0, 1), then
+        x = [b_i] * (P2i.x - identity.x) + identity.x
+          = [b_i] * P2i.x
+        y = [b_i] * (P2i.y - identity.y) + identity.y
+          = [b_i] * (P2i.y - 1) + 1
+    So we get the SharedPoint of each term.
+    """
+    if isinstance(p, Ideal):
+        return SharedIdeal(p.curve)
+
+    terms = []
+    for i in range(len(bs)):
+        p2i = (2**i) * p
+        term = await SharedPoint.create(context,
```

Comment:
Why do we have this `.create` method instead of just the ordinary constructor for `SharedPoint`?  Can we just use the constructor?

Reply:
Currently, it's this way because the existing code I took jubjub from had the `on_curve` check for created points, and you can't use `await` from `__init__` for computing `on_curve`. I think it's reasonable to remove this check for code that should not produce points off the curve.